client/src/JobPaginatorState.js

```diff
@@ -0,0 +1,129 @@
+    });
+    return foundIndex === -1 ? list.length : foundIndex;
+};
+
+const getJob = (category, job) => {
+    const ajax = AjaxHelper.getInstance();
+    return new Promise((resolve /*, reject*/) =>
+        ajax.getJSON(ajax.getJobUrl(category, job), null, (data) => {
+            resolve(data);
+        })
+    );
+};
+
+const getStatus = (category, job) => {
+    const ajax = AjaxHelper.getInstance();
+    return new Promise((resolve) =>
+        ajax.getJSON(ajax.getJobUrl(category, job, "/status"), null, (data) => {
+            resolve(data);
+        })
+    );
+};
+
+const deleteImage = (imageUrl, success, error) => {
+    AjaxHelper.getInstance().ajax({
+        type: "DELETE",
+        url: imageUrl,
+        success,
+        error,
+    });
+};
+
+const putJob = ({ category, job, order, published }, success, error) => {
+    const ajax = AjaxHelper.getInstance();
+    ajax.ajax({
+        type: "PUT",
+        url: ajax.getJobUrl(category, job),
+        contentType: "application/json",
+        data: JSON.stringify({ order, published }),
+        success,
+        error,
+    });
+};
+
+const deletePageValidation = (order) => {
+    if (order.length < 2) {
+        alert("You cannot delete the last page in a job.");
+        return false;
+    }
+    return window.confirm("Are you sure you wish to delete the current page?");
+};
+
+const getAddedPages = (order, pages) => {
+    return pages.reduce((order, page) => {
+        order.splice(findNewPagePosition(page, order), 0, {
+            filename: page,
+            label: null,
+        });
+        return order;
+    }, order);
+};
+
+const getNoneRemovedPages = (order, deletedPages) =>
+    order.filter((o) =>
+        deletedPages.every((page) => {
+            return o.filename !== page;
+        })
+    );
+
+const validatePublish = async (category, job) => {
+    const response = await getStatus(category, job);
+    const {
+        derivatives: { expected, processed },
+    } = response;
+    if (expected > processed) {
+        alert(
+            `Derivative images have not been generated yet. Please` +
+                ` go back to the main menu and hit the "build" button` +
+                ` for this job before publishing it.`
+        );
+        return false;
+    }
+    return !window.confirm(
+        "Are you sure you wish to publish this job? You will not be able to make any further edits."
+    );
+};
+
+const getLabel = (order, imageNumber) => {
+    return typeof order[imageNumber] === "undefined" ? null : order[imageNumber].label;
+};
+
+const countMagicLabels = (order, startAt) => {
+    let count = 0;
+    for (let startIndex = startAt; startIndex < order.length; startIndex++) {
+        if (!getLabel(order, startIndex)) {
+            count++;
+        }
+    }
+    return count;
+};
+
+const validateSave = (order) => {
+    const count = countMagicLabels(order, 0);
+    return count > 0 && !confirmSavedMagicLabels(count);
```

Comment:
The confirmSavedMagicLabels logic is inverted here; I have to hit "cancel" in order to save, and if I hit OK, the save action is canceled. (You can test this by trying to save a job with auto-generated labels).

Reply:
On a possibly-related note, apart from having inverted logic, the "Save" button does seem to work correctly. However, the "Save and Publish" button is more broken. I can't seem to find a way to publish a job at all, and during my attempts, I get this error:

<pre>
Warning: Cannot update a component (`JobPaginator`) while rendering a different component (`Thumbnail`). To locate the bad setState() call inside `Thumbnail`, follow the stack trace as described in https://fb.me/setstate-in-render
    in Thumbnail (at PaginatorList.jsx:18)
    in div (at PaginatorList.jsx:35)
    in PaginatorList (at JobPaginator.jsx:55)
    in div (at JobPaginator.jsx:36)
    in div (at JobPaginator.jsx:32)
    in div (at JobPaginator.jsx:31)
    in JobPaginator (at Routes.jsx:12)
    in JobPaginatorHook (at Routes.jsx:25)
    in Route (at Routes.jsx:24)
    in Switch (at Routes.jsx:17)
    in Routes (at VuDLPrep.jsx:15)
    in Router (created by BrowserRouter)
    in BrowserRouter (at VuDLPrep.jsx:14)
    in div (at VuDLPrep.jsx:10)
    in VuDLPrep (at src/index.js:14)
    in StrictMode (at src/index.js:13)
</pre>

This is followed by some weird, malformed API requests.

Reply:
I modified and made changes.